### ISSUE TITLE
Multimodal: perform session.update from the ToolCall

### ISSUE THREAD

**Issue by mrdrprofuroboros** [livekit/agents#837]
Trying to wrap my head around multimodal agent and openai realtime API :)

I want to steer the conversation by managing the system context and my intuition was that I could supply my agent with a tool, that could send a session.update with new "instructions" during the conversation. I'd like to hack a prototype for that but just decided I should start a discussion here in case you had any approach to this in mind.

so basically, let's say this tool manages a state machine and can be triggered by function calling to make a transition and change the state. once the state has changed, it should load some new system information to the session (new prompts with new instructions). Does that make sense to you? What is the quickest / dirties way to test it right now with Multimodal Agent and what would be a good approach to add this logic to the framework?

**Comment by mrdrprofuroboros** [comment]
right now if I send session.update during the model speech it interrupts speaking, which is .. undesirable

**Comment by vnandan** [comment]
@mrdrprofuroboros  did you ever find a solution? What i am currently trying to work on is that the function call actually blocks the conversation upto some extent. The AI will start to wait for the result before moving on. Our use case requires us to move the conversation forward and not wait for the result. So, the function call simply return a statement like, "Thanks for the response, we're processing it. In the meantime let's discuss..." and pushes the actual calculation to another async function. Then, when the result is ready, we do a session.session_update(instructions="the new thing we want to inject").

Any other approach?

**Comment by psinojiya** [comment]
Hello @vnandan @mrdrprofuroboros did we have any luck on this?

**Comment by vnandan** [comment]
We're working on a  similar problem. session_update was breaking the realtime conversation model's speech mid-sentence. So, we started pushing the updated data/state via a developer/system message. Further, so the conversational model is sensitized via the initial system instruction to use the latest updated state and ignore previous states. This way, the basic system instruction stays the same but the state keeps changing.

**Comment by mrdrprofuroboros** [comment]
@psinojiya yes, we decided not to use realtime APIs because of providers limitations, came up with our own solution from scratch

**Comment by jingxuansunny** [comment]
@mrdrprofuroboros Hi, I am new here, not able to find many informations talking about update instructions during the session. I found the Livekit probive this function to update instructions, but not able to make it work. 

async def update_instructions(self, instructions: str) -> None:
    event_id = utils.shortuuid("instructions_update_")
    # f = asyncio.Future()
    # self._response_futures[event_id] = f
    self.send_event(
        SessionUpdateEvent(
            type="session.update",
            session=session_update_event.Session.model_construct(instructions=instructions),
            event_id=event_id,
        )
    )
Is this how you update the instructions, if not, do you have any suggestions how to use this?

**Comment by vnandan** [comment]
@psinojiya  @mrdrprofuroboros 
I think the official way of doing this is to use 
`function_calls_finished`

We resorted to a more longer way of running the function on a separate thread and passing self to keep the reference, then at the end of the function, call the callback function on the main thread.

We've also moved away from realtime api, it seems less accurate and VoicePipeline gives much more control
